File: app/src/main/java/com/absinthe/kage/media/video/RemoteVideoPlayback.java
package com.absinthe.kage.media.video;

import android.media.session.PlaybackState;
import android.util.Log;

import com.absinthe.kage.connect.proxy.VideoProxy;
import com.absinthe.kage.device.model.VideoInfo;
import com.absinthe.kage.media.LocalMedia;
import com.absinthe.kage.media.Playback;

public class RemoteVideoPlayback implements Playback {
    private static final String TAG = RemoteVideoPlayback.class.getSimpleName();
    private Callback mCallback;
    private VideoProxy mVideoProxy = VideoProxy.getInstance();
    private int mPlayState = PlaybackState.STATE_NONE;

    public RemoteVideoPlayback() {

        mVideoProxy.setOnPlayListener(new VideoProxy.OnPlayListener() {

            @Override
            public void onCurrentPositionChanged(int duration, int position) {

            }

            @Override
            public void onPlayStateChanged(int oldState, int newState) {
                if (newState == VideoProxy.PlayStatue.INVALIDATE) {
                    mPlayState = PlaybackState.STATE_NONE;
                } else if (newState == VideoProxy.PlayStatue.PLAYER_EXIT) {
                    mPlayState = PlaybackState.STATE_STOPPED;
                } else if (newState != VideoProxy.PlayStatue.DISCONNECT) {
                    switch (newState) {
                        case VideoProxy.PlayStatue.STOPPED:
                            mPlayState = PlaybackState.STATE_STOPPED;
                            break;
                        case VideoProxy.PlayStatue.TRANSITIONING:
                            mPlayState = PlaybackState.STATE_BUFFERING;
                            break;
                        case VideoProxy.PlayStatue.PLAYING:
                            mPlayState = PlaybackState.STATE_PLAYING;
                            break;
                        case VideoProxy.PlayStatue.PAUSED:
                            mPlayState = PlaybackState.STATE_PAUSED;
                            break;
                    }
                } else {
                    mPlayState = PlaybackState.STATE_STOPPED;
                }
                handlePlayState(false);
            }
        });
    }

    public void playMedia(LocalMedia localMedia) {
        if (localMedia != null) {
            VideoInfo mediaInfo = new VideoInfo();
            mediaInfo.setUrl(localMedia.getUrl());
            mediaInfo.setTitle(localMedia.getTitle());
            mVideoProxy.play(mediaInfo);
            Log.i(TAG, "playMedia");
            mPlayState = PlaybackState.STATE_BUFFERING;

            if (mCallback != null) {
                mCallback.onMediaMetadataChanged(localMedia);
                mCallback.onPlaybackStateChanged(mPlayState);
                return;
            }
            return;
        }
        mPlayState = PlaybackState.STATE_ERROR;
        if (mCallback != null) {
            mCallback.onPlaybackStateChanged(mPlayState);
        }
    }

    public void play() {
        mPlayState = PlaybackState.STATE_PLAYING;
        handlePlayState(true);
    }

    public void pause() {
        mPlayState = PlaybackState.STATE_PAUSED;
        handlePlayState(true);
    }

    public void seekTo(int position) {
        mVideoProxy.seekTo(position);
    }

    public int getState() {
        return mPlayState;
    }

    public int getDuration() {
        return mVideoProxy.getDuration();
    }

    public int getBufferPosition() {
        return 100;
    }

    public int getCurrentPosition() {
        return mVideoProxy.getCurrentPosition();
    }

    public void setCallback(Callback callback) {
        mCallback = callback;
    }

    public void stop(boolean fromUser) {
        mPlayState = PlaybackState.STATE_STOPPED;
        if (fromUser && mCallback != null) {
            mVideoProxy.stop();
        }
        mVideoProxy.recycle();
    }

    private void handlePlayState(boolean formUser) {
        if (formUser) {
            if (mPlayState == PlaybackState.STATE_PLAYING) {
                mVideoProxy.start();
            } else if (mPlayState == PlaybackState.STATE_PAUSED) {
                mVideoProxy.pause();
            }
        }
        if (mCallback != null) {
            mCallback.onPlaybackStateChanged(mPlayState);
        }
    }
}

File: app/src/main/kotlin/com/absinthe/kage/media/video/RemoteVideoPlayback.kt
package com.absinthe.kage.media.video

import android.media.session.PlaybackState
import com.absinthe.kage.connect.proxy.VideoProxy
import com.absinthe.kage.device.model.VideoInfo
import com.absinthe.kage.media.LocalMedia
import com.absinthe.kage.media.Playback
import timber.log.Timber

class RemoteVideoPlayback : Playback {

    private val mVideoProxy = VideoProxy
    private var mCallback: Playback.Callback? = null

    init {
        mVideoProxy.setOnPlayListener(object : VideoProxy.OnPlayListener {

            override fun onCurrentPositionChanged(duration: Int, position: Int) {}

            override fun onPlayStateChanged(oldState: Int, newState: Int) {
                when {
                    newState == VideoProxy.PlayStatue.INVALIDATE -> {
                        state = PlaybackState.STATE_NONE
                    }
                    newState == VideoProxy.PlayStatue.PLAYER_EXIT -> {
                        state = PlaybackState.STATE_STOPPED
                    }
                    newState != VideoProxy.PlayStatue.DISCONNECT -> {
                        when (newState) {
                            VideoProxy.PlayStatue.STOPPED -> state = PlaybackState.STATE_STOPPED
                            VideoProxy.PlayStatue.TRANSITIONING -> state = PlaybackState.STATE_BUFFERING
                            VideoProxy.PlayStatue.PLAYING -> state = PlaybackState.STATE_PLAYING
                            VideoProxy.PlayStatue.PAUSED -> state = PlaybackState.STATE_PAUSED
                        }
                    }
                    else -> {
                        state = PlaybackState.STATE_STOPPED
                    }
                }
                handlePlayState(false)
            }
        })
    }

    override var state = PlaybackState.STATE_NONE
        private set

    override val duration: Int
        get() = mVideoProxy.duration

    override val bufferPosition: Int
        get() = 100

    override val currentPosition: Int
        get() = mVideoProxy.currentPosition

    override fun playMedia(localMedia: LocalMedia) {
        Timber.i("playMedia")

        val mediaInfo = VideoInfo().apply {
            this.url = localMedia.url
            this.title = localMedia.title
        }
        mVideoProxy.play(mediaInfo)

        state = PlaybackState.STATE_BUFFERING
        mCallback?.onMediaMetadataChanged(localMedia)
        mCallback?.onPlaybackStateChanged(state)
    }

    override fun play() {
        state = PlaybackState.STATE_PLAYING
        handlePlayState(true)
    }

    override fun pause() {
        state = PlaybackState.STATE_PAUSED
        handlePlayState(true)
    }

    override fun seekTo(position: Int) {
        mVideoProxy.seekTo(position)
    }

    override fun setCallback(callback: Playback.Callback) {
        mCallback = callback
    }

    override fun stop(isStop: Boolean) {
        state = PlaybackState.STATE_STOPPED
        if (isStop && mCallback != null) {
            mVideoProxy.stop()
        }
        mVideoProxy.recycle()
    }

    private fun handlePlayState(formUser: Boolean) {
        if (formUser) {
            if (state == PlaybackState.STATE_PLAYING) {
                mVideoProxy.start()
            } else if (state == PlaybackState.STATE_PAUSED) {
                mVideoProxy.pause()
            }
        }
        mCallback?.onPlaybackStateChanged(state)
    }
}